src/dev-app/datepicker/datepicker-demo.html

```diff
@@ -169,3 +169,28 @@ <h2>Datepicker with custom header extending the default header</h2>
     <mat-datepicker #customHeaderNgContentPicker [calendarHeaderComponent]="customHeaderNgContent"></mat-datepicker>
   </mat-form-field>
 </p>
+
+<h2>Range picker</h2>
+<p>
+  <mat-form-field>
+    <mat-label>Enter a date range</mat-label>
+    <mat-date-range-input startPlaceholder="Start date" endPlaceholder="End date"></mat-date-range-input>
```

Comment:
I have a TODO to extract the common logic in a follow-up.

As for projecting the inputs: I was thinking about doing this, but I made them non-projectable, because I feel like we'd need more control over the inputs than usual. E.g. I'm not sure how it would behave if people put `matInput` on them out of habit or they stuck a `FormControl` on it. Also this makes it a bit easier to override styles and people's own styling could conflict with the ones I put in to get the autosize to work. Finally, I just didn't see what else people might want to customize on it, aside from the `placeholder`.

Reply:
I'm not too concerned about the `MatInput` thing. If we're worried about it we could always detect it and show an error. This would definitely be a departure from what our library has tried to do so far (i.e. always expose the native controls). I could imagine someone wanting to make the start or end `readonly` for some reason and then needing to add properties like `isStartReadonly`, `isEndReadonly`, etc. I think its better to just expose the inputs and trust people.

@jelbourn I suspect you will have feels about this

Reply:
The two inputs are just an implementation detail thought. They're technically a single "input".

Reply:
@jelbourn for some reason GitHub put your comment in the wrong order and it's second in the discussion...

Anyway, I agree that it would be better if we allowed users to project their own input, but I don't see a way how we can prevent them from doing weird things and breaking the layout. I also don't see any more properties that we'd need to forward, aside from the `placeholder` since everything, including the input value, is managed by Material. Are there any concrete a11y concerns?

Reply:
I made the draft comment a few hours before sending it, so that's probably why. 

My surface level thoughts were about making sure people can directly use `aria-label`, `aria-labelledby` and `aria-describedby` for a11y, and that `autocomplete` and `autofill` could be relevant. Thinking about it more now, the biggest thing is probably actually forms participation; people will most likely want to treat each input as a separate form control, no?

Reply:
We generate `aria-labelledby` and `aria-describedby` already by pointing them to the relevant elements in the `mat-form-field`. I'm not sure what we should do about `aria-label`.

As for forms participation, I think the idea was for the value to be a `DateRange` object with a `start` and `end` and we'll populate the inputs based on that. Otherwise if we want separate form controls for each field we'll have to go with projecting the inputs.

Reply:
I could see wanting to bind forms to either one. Its nice to be able to bind to the range as a whole so you get validation of the entire range (e.g. verify end comes after start). However I could also see people wanting to bind them individually since they're probably persisted separately instead of as one range object

Reply:
I think that we need to decide on one or the other, because it would be tricky to handle both. It means more work, but I'm leaning towards binding to each field individually, because people can forward values directly from the database, whereas binding the `DateRange` would mean having an intermediate step where they'd have to instantiate the object.

Reply:
That exactly was my thinking too; the data is almost certainly going to be stored in the DB as separate start and end ranges. I also agree that there should only be one way to do it.

Reply:
I am sad that we lose the ability to validate the range as a whole though, by not allowing people to bind a form control to the combined `date-range-input`. Is there some way we could still support that?

I'm also not sure how only have the separate ones will play with `mat-form-field`. I _think_ the form field expects the `MatFormFieldControl` to have a single `FormControl` representing its value (though I'm not 100% sure about that, it's been a while).

Reply:
A `FormGroup` kind of sounds like the right tool for the job, though I'm not sure how that works if people bind the individual inputs with `ngModel`

Reply:
This still works with the `MatFormField` since the `MatFormFieldControl` is the `mat-date-range-input` which then combines the `disabled`, `focused`, `empty` etc. states of the individual inputs so it still works. The demo is using a form field.

Reply:
But it looks like `MatDateRangeInput` is injecting an `NgControl` which doesn't make sense if we don't want people binding a `FormControl` to it. We also can't add validation on the range as a whole, like checking that the start comes before the end.

It seems to me that the ideal solution would be this:
```html
<mat-date-range-input [formGroup]="range">
  <input matStartDate formControlName="start">
  <input matEndDate formControlName="end">
</mat-date-range-input>
```

This gives us a value that makes sense as our `NgControl` (the form group) and allows us to add group-level validation, while still giving the user access to the start and end values separately

Reply:
The `NgControl` is there, because it's required by `MatFormFieldControl`, but it isn't actually used by anything. As for validation, I think that we can validate them separately in the context of the entire range. E.g. if the start is after the end, we can mark the end as being invalid.

Reply:
That's what I mean, just supplying a fake one might make some things not hook up right. `MatFormFieldControl` requires it for a reason: https://github.com/angular/components/blob/d550e4ef97a83534f9b9b0914dc27a3ad46734ec/src/material/form-field/form-field.ts#L319

Your suggested method of validation might work for validations we want to add automatically, but I don't think it'll be a nice experience for users that want to add custom validation (e.g. maximum range length, minimum range length, etc)

Reply:
I don't think the `NgControl` always being null is a problem here. It's just there to ensure that changes are detected correctly which we can compensate for by emitting to `stateChanges` when the range is changed.

As for validation, there are already common cases where people validate one field based on another (e.g. password confirmations) so I'm not sure that most cases would be an issue. I see your point that validating the range length would be tricky, but I'm not sure how common that would be.

My problem with having the `ControlValueAccessor` on the wrapper is that it introduces two extra steps any time the component is consumed:

1. Wrapping the value in a `DateRange` when it comes in from the backend. 
2. Unwrapping the `DateRange` before passing it back to the backend. 

This feels like it would be cumbersome, in addition to the fact that people would be forced to use our own data format.

Reply:
I don't think that's actually true if the recommended way to use it is with a `FormGroup`. That way each input gets its own control inside the form group, but you still have the ability to add validation at the group level

Reply:
Changed based on our discussion on Slack: the `NgControl` for the group will be the `ControlContainer`. It's currently being cast as `any` until #18206 lands.